Handling flaky tests

For roughly the past month, I have collected data on flaky CI tests. The initial idea was to mark them as [flaky](https://github.com/box/flaky), but as per [pytest's docs on flaky tests](https://docs.pytest.org/en/stable/explanation/flaky.html), that should never be a long-term solution. Instead, tests should be (randomly) re-ordered, re-written for more atomic assertions, or split up into different groups to find the root cause of the flaky behavior and eliminate it. We will have to see when time permits this. For now, we could mark them as flaky to save us from re-running them manually. 
Here are the flaky tests of this repository I gathered so far:

# Flaky tests

## Auxiliary

Some error message are common and so long that they would render the table even more complex than it already is. They are included here and referenced by their names in the table below.

### Notebook cell timeout prepare reporter {#notebook-cell-timeout-prepare-reporter}
```python
nbclient.exceptions.CellTimeoutError: A cell timed out while it was being executed, after 10 seconds.
The message was: Cell execution timed out.
Here is a preview of the cell contents:
-------------------
['# Prepare a Reporter based on the solved scenario', 'ctax_rep = Reporter.from_scenario(ctax_scen)', 'prepare_plots(ctax_rep)', '', '# Run the same reporting on both the base and modified scenarios']
...
['    r.get("plot new capacity")', '    r.get("plot capacity")', '    r.get("plot activity")', '    r.set_filters(t=None, c=["light", "other_electricity"])', '    r.get("plot prices")']
-------------------
```

### Notebook cell timeout import packages {#notebook-cell-timeout-import-packages}
```python
nbclient.exceptions.CellTimeoutError: A cell timed out while it was being executed, after 10 seconds.
The message was: Cell execution timed out.
Here is a preview of the cell contents:
-------------------
# Importing required software packages
import pandas as pd
import ixmp
import message_ix

from message_ix.utils import make_df

%matplotlib inline
-------------------
```


Names are shortened to `message_ix/tests` as the starting directory. Grouping the tests by names shows commonalities between them. Numerical errors seem to appear exclusively for macos, while ubuntu only sports errors in testing the tutorials, almost all of which are similar to one another. Windows appears only once when it was unable to connect to the database (these two errors are related, see also ). 


| Test name    | Error message | Runners (# of occurrences if > 1) |
|--------------|:-------------:|---------------------------:|
| test_feature_bound_activity_shares.py:: test_add_bound_activity_up_all_modes |  `ixmp.model.base.ModelError: GAMS errored with return code 3: There was an execution error` |        macos-latest-py3.7 (2), macos-latest-py3.8, macos-latest-py3.10 (2), macos-latest-py3.11 (2) |
| test_feature_bound_activity_shares.py:: test_add_bound_activity_up_all_modes | `assert False where False = <function isclose at 0x10c08edd0>(300.0, 332.5) where <function isclose at 0x10c08edd0> = np.isclose` | macos-latest-py3.7, macos-latest-py3.8, macos-latest-py3.9, macos-latest-py3.10 (3) |
| test_feature_bound_activity_shares.py:: test_share_commodity_lo | `assert False where False = <function isclose at 0x11054bf60>(0.15384615384615385, 0.0) where <function isclose at 0x11054bf60> = np.isclose` | macos-latest-py3.11 |
| test_feature_bound_activity_shares.py:: test_add_share_mode_lo | `assert False where False = <function isclose at 0x107278860>(0.5416666666666666, 0.525) where <function isclose at 0x107278860> = np.isclose` | macos-latest-py3.11 | 
|------------------|------------------|------------------|
|------------------|------------------|------------------|
| test_feature_price_emission.py:: test_cumulative_equidistant | `at.ac.iiasa.ixmp.exceptions.at.ac.iiasa.ixmp.exceptions.IxException: at.ac.iiasa.ixmp.exceptions.IxException: The index set 'year' does not have an element '2025'!` |          macos-latest-py3.7 |
| test_feature_price_emission.py:: test_per_period_variable_periodlength | `at.ac.iiasa.ixmp.exceptions.at.ac.iiasa.ixmp.exceptions.IxException: at.ac.iiasa.ixmp.exceptions.IxException: The index set 'year' does not have an element '2050'!` | macos-latest-py3.10 |
| test_feature_price_emission.py:: test_custom_type_variable_periodlength | ```AssertionError: Not equal to tolerance rtol=1e-07, atol=0 (shapes (4,), (3,) mismatch)  x: array([1., 1., 1., 1.])  y: array([1., 1.276282, 2.078928])``` | macos-latest-py3.7, macos-latest-py3.8 |
| test_feature_price_emission.py:: test_cumulative_variable_periodlength | `AssertionError: Not equal to tolerance rtol=1e-07, atol=0 Mismatched elements: 3 / 4 (75%) Max absolute difference: 1.65329771 Max relative difference: 0.62311052 x: array([1., 1., 1., 1.])  y: array([1., 1.276282, 1.628895, 2.653298])` | macos-latest-py3.7, macos-latest-py3.8, macos-latest-py3.10, macos-latest-py3.11 |
| test_feature_price_emission.py:: test_per_period_variable_periodlength | `AssertionError: Not equal to tolerance rtol=1e-07, atol=0 (shapes (4,), (3,) mismatch) x: array([1., 1., 1., 1.]) y: array([1.      , 1.276282, 2.078928])` | macos-latest-py3.9 |
| test_feature_price_emission.py:: test_per_period_variable_periodlength | `assert 0.0 > 0` | macos-latest-py3.9 |
|------------------|------------------|------------------|
|------------------|------------------|------------------|
| test_macro.py:: test_calc[_demand-allclose-expected6] | `ixmp.model.base.ModelError: GAMS errored with return code 3: There was an execution error` | macos-latest-py3.8 |
| test_macro.py:: test_calc[_gdp0-equal-expected2] | `ixmp.model.base.ModelError: GAMS errored with return code 2: There was a compilation error` | macos-latest-py3.7, macos-latest-py3.9 |
| test_macro.py:: test_config | `ixmp.model.base.ModelError: GAMS errored with return code 2: There was a compilation error` | macos-latest-py3.11 |
| test_macro.py:: test_calc_valid_data_file | `ixmp.model.base.ModelError: GAMS errored with return code 2: There was a compilation error` | macos-latest-py3.10 |
| test_macro.py:: test_add_model_data | `at.ac.iiasa.ixmp.exceptions.at.ac.iiasa.ixmp.exceptions.IxException: at.ac.iiasa.ixmp.exceptions.IxException: The index set 'node' does not have an element 'seattle'!` | macos-latest-py3.9 (2) |
| test_macro.py:: test_calc[_aconst-allclose-expected8] | `com.gams.api.com.gams.api.GAMSException: com.gams.api.GAMSException: Error opening GDX file: Expected data marker (SYMBOL_2) not found in GDX file` | macos-latest-py3.11 |
|------------------|------------------|------------------|
|------------------|------------------|------------------|
| test_legacy_version.py:: test_solve_legacy_scenario | `ixmp.model.base.ModelError: GAMS errored with return code 3: There was an execution error` | macos-latest-py3.9 |
| test_legacy_version.py:: test_solve_legacy_scenario | `at.ac.iiasa.ixmp.exceptions.at.ac.iiasa.ixmp.exceptions.IxException: at.ac.iiasa.ixmp.exceptions.IxException: The index set 'year' does not have an element '1963'!` | macos-latest-py3.7, macos-latest-py3.9 (2), macos-latest-py3.10, macos-latest-py3.11  | 
|------------------|------------------|------------------|
|------------------|------------------|------------------|
| test_integration.py:: test_multi_db_run | `at.ac.iiasa.ixmp.exceptions.at.ac.iiasa.ixmp.exceptions.IxException: at.ac.iiasa.ixmp.exceptions.IxException: The index set 'year' does not have an element '2010'!` | macos-latest-py3.9 |
|------------------|------------------|------------------|
|------------------|------------------|------------------|
| test_tutorials.py:: test_tutorial[austria_multiple_policies-answers]@ at0 | [Notebook cell timeout prepare reporter](#notebook-cell-timeout-prepare-reporter) | ubuntu-latest-py3.10 |
| test_tutorials.py:: test_tutorial[westeros_addon_technologies]@ w0 | [Notebook cell timeout import packages](#notebook-cell-timeout-import-packages) | ubuntu-latest-py3.10 |
| test_tutorials.py:: test_tutorial[westeros_multinode]@ w0 | [Notebook cell timeout import packages](#notebook-cell-timeout-import-packages) with additional `from itertools import product` | ubuntu-latest-py3.10 |
| test_tutorials.py:: test_tutorial[westeros_share_constraint]@ w0 | [Notebook cell timeout import packages](#notebook-cell-timeout-import-packages) w/o import from `message_ix.util` | ubuntu-latest-py3.10,  ubuntu-latest-py3.11 |
| test_tutorials.py:: test_tutorial[westeros_soft_constraints]@ w0 | [Notebook cell timeout import packages](#notebook-cell-timeout-import-packages) w/o import from `message_ix.util` | ubuntu-latest-py3.10, ubuntu-latest-py3.11 |


See also iiasa/ixmp#489 and iiasa/message-ix-models#112.

--- a/message_ix/tests/test_feature_bound_activity_shares.py
+++ b/message_ix/tests/test_feature_bound_activity_shares.py
@@ -1,8 +1,20 @@
+import os
+import platform
+
 import numpy as np
 import pandas as pd
+import pytest
 
 from message_ix import Scenario
 from message_ix.testing import SCENARIO
+
+FLAKY = pytest.mark.flaky(
+    reruns=5,
+    rerun_delay=2,
+    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Darwin",
+    reason="Flaky; see iiasa/message_ix#731",
+)
+
 
 # First model year of the Dantzig scenario
 _year = 1963
@@ -45,6 +57,7 @@ def test_add_bound_activity_up(message_test_mp):
     assert new_obj >= orig_obj
 
 
+@FLAKY
 def test_add_bound_activity_up_all_modes(message_test_mp):
     # This test specifically has two solutions for which the `OBJ` function is the same
     # therefore the lpmethod must be set to "2".
@@ -228,7 +241,8 @@ def test_commodity_share_up(message_test_mp):
     assert new_obj >= orig_obj
 
 
-def test_share_commodity_lo(message_test_mp):
+@FLAKY
+def test_commodity_share_lo(message_test_mp):
     scen = Scenario(message_test_mp, **SCENARIO["dantzig"]).clone()
     scen.solve(quiet=True)
 
@@ -344,6 +358,7 @@ def test_add_share_mode_up(message_test_mp):
     assert new_obj >= orig_obj
 
 
+@FLAKY
 def test_add_share_mode_lo(message_test_mp):
     scen = Scenario(message_test_mp, **SCENARIO["dantzig"]).clone()
     scen.solve(quiet=True)

--- a/message_ix/tests/test_feature_price_emission.py
+++ b/message_ix/tests/test_feature_price_emission.py
@@ -1,6 +1,18 @@
+import os
+import platform
+
 import numpy.testing as npt
+import pytest
 
 from message_ix import Scenario
+
+FLAKY = pytest.mark.flaky(
+    reruns=5,
+    rerun_delay=2,
+    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Darwin",
+    reason="Flaky; see iiasa/message_ix#731",
+)
+
 
 MODEL = "test_emissions_price"
 
@@ -74,6 +86,7 @@ def test_no_constraint(test_mp):
     assert scen.var("PRICE_EMISSION").empty
 
 
+@FLAKY
 def test_cumulative_equidistant(test_mp):
     scen = Scenario(test_mp, MODEL, "cum_equidistant", version="new")
     years = [2020, 2030, 2040]
@@ -110,6 +123,7 @@ def test_per_period_equidistant(test_mp):
     npt.assert_allclose(scen.var("PRICE_EMISSION")["lvl"], [1] * 3)
 
 
+@FLAKY
 def test_cumulative_variable_periodlength(test_mp):
     scen = Scenario(test_mp, MODEL, "cum_equidistant", version="new")
     years = [2020, 2025, 2030, 2040]
@@ -128,6 +142,7 @@ def test_cumulative_variable_periodlength(test_mp):
     npt.assert_allclose(obs, [1.05 ** (y - years[0]) for y in years])
 
 
+@FLAKY
 def test_per_period_variable_periodlength(test_mp):
     scen = Scenario(test_mp, MODEL, "cum_equidistant", version="new")
     years = [2020, 2025, 2030, 2040]
@@ -146,6 +161,7 @@ def test_per_period_variable_periodlength(test_mp):
     npt.assert_allclose(scen.var("PRICE_EMISSION")["lvl"].values, [1] * 4)
 
 
+@FLAKY
 def test_custom_type_variable_periodlength(test_mp):
     scen = Scenario(test_mp, MODEL, "cum_equidistant", version="new")
     years = [2020, 2025, 2030, 2040, 2050]

--- a/message_ix/tests/test_integration.py
+++ b/message_ix/tests/test_integration.py
@@ -1,3 +1,6 @@
+import os
+import platform
+
 import numpy as np
 import pytest
 from ixmp import Platform
@@ -84,6 +87,12 @@ def assert_multi_db(mp1, mp2):
     assert_frame_equal(scenario_list(mp1), scenario_list(mp2))
 
 
+@pytest.mark.flaky(
+    reruns=5,
+    rerun_delay=2,
+    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Darwin",
+    reason="Flaky; see iiasa/message_ix#731",
+)
 def test_multi_db_run(tmpdir):
     # create a new instance of the transport problem and solve it
     mp1 = Platform(driver="hsqldb", path=tmpdir / "mp1")

--- a/message_ix/tests/test_legacy_version.py
+++ b/message_ix/tests/test_legacy_version.py
@@ -1,10 +1,20 @@
+import os
+import platform
+
 import numpy as np
+import pytest
 from ixmp import Platform
 from ixmp.testing import create_test_platform
 
 from message_ix import Scenario
 
 
+@pytest.mark.flaky(
+    reruns=5,
+    rerun_delay=2,
+    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Darwin",
+    reason="Flaky; see iiasa/message_ix#731",
+)
 def test_solve_legacy_scenario(tmp_path, test_data_path):
     db_path = create_test_platform(tmp_path, test_data_path, "legacy")
     mp = Platform(backend="jdbc", driver="hsqldb", path=db_path)

--- a/message_ix/tests/test_macro.py
+++ b/message_ix/tests/test_macro.py
@@ -1,3 +1,5 @@
+import os
+import platform
 from pathlib import Path
 
 import numpy as np
@@ -9,6 +11,13 @@ from message_ix import Scenario, macro
 from message_ix.models import MACRO
 from message_ix.reporting import Quantity
 from message_ix.testing import SCENARIO, make_westeros
+
+FLAKY = pytest.mark.flaky(
+    reruns=5,
+    rerun_delay=2,
+    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Darwin",
+    reason="Flaky; see iiasa/message_ix#731",
+)
 
 # Fixtures
 
@@ -51,6 +60,7 @@ def westeros_not_solved(_ws):
 # Tests
 
 
+@FLAKY
 def test_calc_valid_data_file(westeros_solved, w_data_path):
     c = macro.prepare_computer(westeros_solved, data=w_data_path)
     c.get("check all")
@@ -97,6 +107,7 @@ def test_calc_no_solution(westeros_not_solved, w_data_path):
         macro.prepare_computer(s, data=w_data_path)
 
 
+@FLAKY
 def test_config(westeros_solved, w_data_path):
     c = macro.prepare_computer(westeros_solved, data=w_data_path)
     assert "config::macro" in c.graph
@@ -172,6 +183,7 @@ def test_calc_data_missing_datapoint(westeros_solved, w_data):
 #
 
 
+@FLAKY
 @pytest.mark.parametrize(
     "key, test, expected",
     (
@@ -236,6 +248,7 @@ def test_init(message_test_mp):
     assert "COST_ACCOUNTING_NODAL" in scen.equ_list()
 
 
+@FLAKY
 def test_add_model_data(westeros_solved, w_data_path):
     base = westeros_solved
     clone = base.clone("foo", "bar", keep_solution=False)

--- a/message_ix/tests/test_tutorials.py
+++ b/message_ix/tests/test_tutorials.py
@@ -30,7 +30,7 @@ MARK = [
 
 # Affects all tests in the file
 pytestmark = pytest.mark.flaky(
-    reruns=3,
+    reruns=5,
     rerun_delay=2,
     condition=GHA,
     reason="Flaky; fails occasionally on GitHub Actions runners",
@@ -121,7 +121,7 @@ def nb_path(request, tutorial_path):
 
 def default_args():
     """Default arguments for :func:`.run_notebook."""
-    if GHA and sys.platform in ("darwin", "win32"):
+    if GHA:
         # Use a longer timeout
         return dict(timeout=30)
     else:

